examples/shell/shell_common/cmd_ping.cpp

```diff
@@ -0,0 +1,436 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <core/CHIPCore.h>
+#include <messaging/ExchangeMgr.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <protocols/echo/Echo.h>
+#include <shell/shell.h>
+#include <support/CodeUtils.h>
+#include <support/ErrorStr.h>
+#include <system/SystemPacketBuffer.h>
+#include <transport/PASESession.h>
+#include <transport/SecureSessionMgr.h>
+#include <transport/raw/TCP.h>
+#include <transport/raw/UDP.h>
+
+#include <ChipShellCollection.h>
+
+using namespace chip;
+using namespace Shell;
+using namespace Logging;
+
+#define ECHO_CLIENT_PORT (CHIP_PORT + 1)
+
+constexpr size_t kMaxTcpActiveConnectionCount = 4;
+constexpr size_t kMaxTcpPendingPackets        = 4;
+constexpr size_t kNetworkSleepTimeMsecs       = (100 * 1000);
+
+namespace {
+
+constexpr Transport::AdminId gAdminId = 0;
+
+// The echo client object.
+Protocols::Echo::EchoClient gEchoClient;
+
+TransportMgr<Transport::UDP> gUDPManager;
+TransportMgr<Transport::TCP<kMaxTcpActiveConnectionCount, kMaxTcpPendingPackets>> gTCPManager;
+Messaging::ExchangeManager gExchangeManager;
+SecureSessionMgr gSessionManager;
+Inet::IPAddress gDestAddr;
+
+// Max value for the number of echo requests sent.
+uint32_t gMaxEchoCount = 3;
+
+// The CHIP Echo interval time in milliseconds.
+uint32_t gEchoInterval = 1000;
+
+// The last time a echo request was attempted to be sent.
+uint64_t gLastEchoTime = 0;
+
+// Count of the number of echo requests sent.
+uint64_t gEchoCount = 0;
+
+// Count of the number of echo responses received.
+uint64_t gEchoRespCount = 0;
+
+// True, if the echo client is waiting for an echo response
+// after sending an echo request, false otherwise.
+bool gWaitingForEchoResp = false;
+
+bool gUseTCP = false;
+
+bool gUseCRMP = true;
+
+void ResetGlobalState()
+{
+    gMaxEchoCount       = 3;
+    gEchoInterval       = 1000;
+    gLastEchoTime       = 0;
+    gEchoCount          = 0;
+    gEchoRespCount      = 0;
+    gWaitingForEchoResp = false;
+    gUseTCP             = false;
+    gUseCRMP            = true;
+}
+
+bool EchoIntervalExpired(void)
+{
+    uint64_t now = System::Timer::GetCurrentEpoch();
+
+    return (now >= gLastEchoTime + gEchoInterval);
+}
+
+CHIP_ERROR SendEchoRequest(streamer_t * stream)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Messaging::SendFlags sendFlags;
+    const char kRequestFormat[] = "Echo Message %" PRIu64 "\n";
+    char requestData[(sizeof kRequestFormat) + 20 /* uint64_t decimal digits */];
```

Comment:
should we use a constant for the 20? like `kDecimalDigitsForUint64`

Reply:
acked